src/System.Net.Security/src/System/Net/SecureProtocols/SslStreamInternal.cs

```diff
@@ -40,52 +46,134 @@ internal SslStreamInternal(SslState sslState)
                 }
 
                 FreeReadBuffer();
             }
-            if (_PinnableOutputBuffer != null)
+            if (_pinnableOutputBuffer != null)
             {
                 if (PinnableBufferCacheEventSource.Log.IsEnabled())
                 {
-                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Write Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_PinnableOutputBuffer));
+                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Write Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_pinnableOutputBuffer));
                 }
 
-                s_PinnableWriteBufferCache.FreeBuffer(_PinnableOutputBuffer);
+                s_PinnableWriteBufferCache.FreeBuffer(_pinnableOutputBuffer);
             }
         }
 
         internal int Read(byte[] buffer, int offset, int count)
         {
-            return ProcessRead(buffer, offset, count);
+            return ProcessRead(buffer, offset, count, null);
         }
 
         internal void Write(byte[] buffer, int offset, int count)
         {
-            ProcessWrite(buffer, offset, count);
+            ProcessWrite(buffer, offset, count, null);
+        }
+
+        internal IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback asyncCallback, object asyncState)
+        {
+            BufferAsyncResult bufferResult = new BufferAsyncResult(this, buffer, offset, count, asyncState, asyncCallback);
+            AsyncProtocolRequest asyncRequest = new AsyncProtocolRequest(bufferResult);
+            ProcessRead(buffer, offset, count, asyncRequest);
+            return bufferResult;
+        }
+
+        internal int EndRead(IAsyncResult asyncResult)
+        {
+            if (asyncResult == null)
+            {
+                throw new ArgumentNullException("asyncResult");
+            }
+
+            BufferAsyncResult bufferResult = asyncResult as BufferAsyncResult;
+            if (bufferResult == null)
+            {
+                throw new ArgumentException(SR.Format(SR.net_io_async_result, asyncResult.GetType().FullName), "asyncResult");
+            }
+
+            if (Interlocked.Exchange(ref _nestedRead, 0) == 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndRead"));
+            }
+
+            // No "artificial" timeouts implemented so far, InnerStream controls timeout.
+            bufferResult.InternalWaitForCompletion();
+
+            if (bufferResult.Result is Exception)
+            {
+                if (bufferResult.Result is IOException)
+                {
+                    throw (Exception)bufferResult.Result;
```

Comment:
Unrelated to this PR, but I noticed this pattern when looking at the other SslStream race condition issue, and we should probably look at addressing it separately...

There are a bunch of places in the code where an Exception is caught and then later thrown.  This kind of re-throwing ends up clobbering the original call stack and Watson bucket stored in the Exception, making it more difficult for someone looking at the exception to understand where it came from.  To address this, ExceptionDispatchInfo was added in .NET 4.5, and is now used in places like Task that have this exact issue.  I'd suggest we look at using ExceptionDispatchInfo in all of these places.  Basically, instead of just storing an Exception, you'd store the result of ```ExceptionDispatchInfo.Capture(caughtException)```, and then instead of using ```throw _caughtException;```, you'd do ```_caughtExceptionDispatchInfo.Throw();```.  ExceptionDispatchInfo will preserve the original information and append the new throw location's call stack rather than using it to overwrite the original.